graphbolt/src/cuda/unique_and_compact_impl.cu

```diff
@@ -32,6 +33,23 @@ struct EqualityFunc {
   }
 };
 
+#define DefineReductionFunction(reduce_fn, name)                              \
+  template <typename scalar_iterator_t>                                       \
+  auto name(const scalar_iterator_t input, int64_t size) {                    \
+    auto allocator = cuda::GetAllocator();                                    \
+    auto stream = cuda::GetCurrentStream();                                   \
+    using scalar_t = std::remove_reference_t<decltype(input[0])>;             \
+    cuda::CopyScalar<scalar_t> result;                                        \
+    size_t workspace_size = 0;                                                \
+    reduce_fn(nullptr, workspace_size, input, result.get(), size, stream);    \
+    auto temp = allocator.AllocateStorage<char>(workspace_size);              \
+    reduce_fn(temp.get(), workspace_size, input, result.get(), size, stream); \
```

Comment:
A strange thing is `CopyScalar` here only works as a storage but copy nothing, please correct if misunderstanding.

Reply:
Yes, in this case we can use it as storage as the API takes a scalar as input. Other times, such as copying the last element of indptr to get the num edges, it performs a copy when constructed with a device pointer. It also has the facilities to synchronize the stream for the copy. However, we use the least functionality we can depending on the situation to maximize performance.

Reply:
The copy happens implicitly as the underlying kernel directly writes the result into the CPU RAM, since the storage is pinned, we can do that.